## Alex Khvatov Homework #8

In [1]:
#!wget https://github.com/SVizor42/ML_Zoomcamp/releases/download/straight-curly-data/data.zip

In [2]:
#!unzip data.zip

In [3]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import load_img
#from tensorflow.keras.applications.xception import preprocess_input

In [4]:
import numpy as np

SEED = 42
np.random.seed(SEED)
tf.random.set_seed(SEED)

For this homework we will use Convolutional Neural Network (CNN). Like in the lectures, we'll use Keras.

* The shape for input should be (200, 200, 3)
* Next, create a convolutional layer (Conv2D):
    * Use 32 filters
    * Kernel size should be (3, 3) (that's the size of the filter)
    * Use 'relu' as activation
* Reduce the size of the feature map with max pooling (MaxPooling2D)
    * Set the pooling size to (2, 2)
* Turn the multi-dimensional result into vectors using a Flatten layer
* Next, add a Dense layer with 64 neurons and 'relu' activation
* Finally, create the Dense layer with 1 neuron - this will be the output
    * The output layer should have an activation - use the appropriate activation for the binary classification case

In [5]:
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import MaxPooling2D
from tensorflow.keras.layers import Flatten
from tensorflow.keras.layers import Dense


In [6]:
def make_model():
    inputs = keras.Input(shape=(200, 200, 3))

    
    layer1 = Conv2D(filters = 32, kernel_size = (3,3), activation='relu')(inputs)
    
    layer2 = MaxPooling2D(pool_size=(2,2))(layer1)
    layer3 = Flatten()(layer2)
    layer4 = Dense(64,activation='relu')(layer3)

    outputs = Dense(1, activation='sigmoid')(layer4)
    optimizer = keras.optimizers.SGD(learning_rate=0.002, momentum=0.8)
    loss = keras.losses.BinaryCrossentropy()

    

    model = keras.Model(inputs, outputs)
    
    model.compile(
        optimizer=optimizer, 
        loss=loss, 
        metrics=['accuracy']
    )
    return model

### Question 1
Since we have a binary classification problem, what is the best loss function for us?


__Answer__: BinaryCrossentropy


### Question 2
What's the total number of parameters of the model? You can use the __summary__ method for that

In [7]:
model = make_model()

In [8]:
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 200, 200, 3)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d (Conv2D)                 │ (None, 198, 198, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 99, 99, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 313632)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 64)             │    20,072,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 20,073,473 (76.57 MB)

 Trainable params: 20,073,473 (76.57 MB)

 Non-trainable params: 0 (0.00 B)

__Answer__: 20,073,473

### Generators and Training
For the next two questions, use the following data generator for both train and test sets:

_ImageDataGenerator(rescale=1./255)_


In [9]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

train_gen = ImageDataGenerator(rescale=1./255)
train_ds=train_gen.flow_from_directory('./data/train', target_size=(200,200), class_mode='binary',batch_size=20, shuffle=True)

test_gen = ImageDataGenerator(rescale=1./255)
test_ds=test_gen.flow_from_directory('./data/test', target_size=(200,200), class_mode='binary',batch_size=20, shuffle=False)


Found 800 images belonging to 2 classes.
Found 213 images belonging to 2 classes.


In [10]:
history = model.fit(train_ds, epochs=10, validation_data=test_ds)

/Users/AKhvatov/.local/share/virtualenvs/machine-learning-zoomcamp-homework-bLdGrrwV/lib/python3.11/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 6s 128ms/step - accuracy: 0.5425 - loss: 0.7398 - val_accuracy: 0.6291 - val_loss: 0.6488
Epoch 2/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 124ms/step - accuracy: 0.6456 - loss: 0.6398 - val_accuracy: 0.6150 - val_loss: 0.6504
Epoch 3/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 117ms/step - accuracy: 0.6853 - loss: 0.6076 - val_accuracy: 0.6573 - val_loss: 0.6234
Epoch 4/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 114ms/step - accuracy: 0.6485 - loss: 0.6040 - val_accuracy: 0.6526 - val_loss: 0.6166
Epoch 5/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 114ms/step - accuracy: 0.6971 - loss: 0.5763 - val_accuracy: 0.6761 - val_loss: 0.6147
Epoch 6/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 111ms/step - accuracy: 0.6751 - loss: 0.5780 - val_accuracy: 0.6291 - val_loss: 0.6423
Epoch 7/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 113ms/step - accuracy: 0.6947 - loss: 0.5573 - val_accuracy: 0.6573 - val_loss: 0.6115
Epoch 8/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 5s 114ms/step - accuracy: 0.7152 - loss: 0.5456 - val_accuracy: 0.

### Question 3
What is the median of training accuracy for all the epochs for this model?

In [11]:
np.median(history.history['accuracy'])

np.float64(0.6943750083446503)

__Answer__:0.69 <-- when using 'sigmoid' activation

## Question 4
What is the standard deviation of training loss for all the epochs for this model?

In [12]:
np.std(history.history['loss'])

np.float64(0.05538256064110054)

__Answer__: 0.068

### Data Augmentation
For the next two questions, we'll generate more data using data augmentations.

Add the following augmentations to your training data generator:

* rotation_range=50,
* width_shift_range=0.1,
* height_shift_range=0.1,
* zoom_range=0.1,
* horizontal_flip=True,
* fill_mode='nearest'

In [13]:
train_gen = ImageDataGenerator(rescale=1./255, 
                               rotation_range=50,
                               width_shift_range=0.1,
                               height_shift_range=0.1,
                               zoom_range=0.1,
                               horizontal_flip=True,
                               fill_mode='nearest')
train_ds=train_gen.flow_from_directory('./data/train', target_size=(200,200), class_mode='binary',batch_size=20, shuffle=True)

Found 800 images belonging to 2 classes.


### Question 5
Let's train our model for 10 more epochs using the same code as previously.

_Note: make sure you don't re-create the model - we want to continue training the model we already started training._

What is the _mean of test loss_ for all the epochs for the model trained with augmentations?

In [14]:
#retrain the model for 10 more epochs -- totalling 20

#making sure not to re-create the model we already trained previously

In [15]:
history = model.fit(train_ds, epochs=20, validation_data=test_ds)

Epoch 1/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 172ms/step - accuracy: 0.6092 - loss: 0.6458 - val_accuracy: 0.6620 - val_loss: 0.5777
Epoch 2/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 170ms/step - accuracy: 0.6466 - loss: 0.6167 - val_accuracy: 0.6526 - val_loss: 0.5900
Epoch 3/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 167ms/step - accuracy: 0.6503 - loss: 0.6118 - val_accuracy: 0.6714 - val_loss: 0.6184
Epoch 4/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 169ms/step - accuracy: 0.6510 - loss: 0.6114 - val_accuracy: 0.6808 - val_loss: 0.5682
Epoch 5/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 171ms/step - accuracy: 0.6905 - loss: 0.5823 - val_accuracy: 0.7089 - val_loss: 0.5957
Epoch 6/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 174ms/step - accuracy: 0.7179 - loss: 0.5739 - val_accuracy: 0.6854 - val_loss: 0.5596
Epoch 7/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 170ms/step - accuracy: 0.6749 - loss: 0.5926 - val_accuracy: 0.6667 - val_loss: 0.5559
Epoch 8/20
40/40 ━━━━━━━━━━━━━━━━━━━━ 7s 166ms/step - accuracy: 0.6984 - loss: 0.5816 - val_accuracy: 0.

In [16]:
np.mean(history.history['val_loss'])

np.float64(0.5596464544534683)

__Answer__: 0.56

### Question 6
What's the average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

In [17]:
np.average(history.history['val_accuracy'])

np.float64(0.7016431897878647)

__Answer__: 0.70